File: day2_adlc/evals/check.js
#!/usr/bin/env node
'use strict';

// Layer 1: code-based checks. Deterministic, free, runs on every output.
// Catches structural failures. Cannot catch a fluent, valid, WRONG summary.

const fs = require('node:fs');
const path = require('node:path');

const SENTIMENTS = ['positive', 'neutral', 'negative', 'mixed'];
const CATEGORIES = ['billing', 'bug', 'feature_request', 'question', 'other'];
const SUMMARY_MIN = 20;
const SUMMARY_MAX = 200;

const outputsDir = path.join(__dirname, 'outputs');
const ticketsDir = path.join(__dirname, 'tickets');

function checkFile(file) {
  const problems = [];
  const raw = fs.readFileSync(path.join(outputsDir, file), 'utf8');

  let data;
  try {
    data = JSON.parse(raw);
  } catch (err) {
    return [`invalid JSON: ${err.message}`];
  }

  for (const field of ['ticket', 'sentiment', 'category', 'summary']) {
    if (typeof data[field] !== 'string') problems.push(`missing or non-string field: ${field}`);
  }
  if (problems.length) return problems;

  if (!fs.existsSync(path.join(ticketsDir, data.ticket)))
    problems.push(`references unknown ticket: ${data.ticket}`);
  if (!SENTIMENTS.includes(data.sentiment))
    problems.push(`sentiment "${data.sentiment}" not in [${SENTIMENTS.join(', ')}]`);
  if (!CATEGORIES.includes(data.category))
    problems.push(`category "${data.category}" not in [${CATEGORIES.join(', ')}]`);
  if (data.summary.length < SUMMARY_MIN || data.summary.length > SUMMARY_MAX)
    problems.push(`summary length ${data.summary.length} outside ${SUMMARY_MIN}-${SUMMARY_MAX}`);

  return problems;
}

const files = fs.readdirSync(outputsDir).filter((f) => f.endsWith('.json')).sort();
let failures = 0;

for (const file of files) {
  const problems = checkFile(file);
  if (problems.length === 0) {
    console.log(`  PASS  ${file}`);
  } else {
    failures++;
    console.log(`  FAIL  ${file}`);
    for (const p of problems) console.log(`        - ${p}`);
  }
}

console.log(`\n${files.length - failures}/${files.length} outputs pass code checks`);
console.log('Remember: passing here only means structurally valid — not truthful.');
process.exit(failures > 0 ? 1 : 0);

File: day2_adlc/evals/check.py
#!/usr/bin/env python3
"""Layer 1: code-based checks. Deterministic, free, runs on every output.

Catches structural failures. Cannot catch a fluent, valid, WRONG summary.
"""

import json
import sys
from pathlib import Path

SENTIMENTS = {"positive", "neutral", "negative", "mixed"}
CATEGORIES = {"billing", "bug", "feature_request", "question", "other"}
SUMMARY_MIN = 20
SUMMARY_MAX = 200

HERE = Path(__file__).parent
OUTPUTS_DIR = HERE / "outputs"
TICKETS_DIR = HERE / "tickets"


def check_file(path: Path) -> list[str]:
    problems = []
    raw = path.read_text()

    try:
        data = json.loads(raw)
    except json.JSONDecodeError as err:
        return [f"invalid JSON: {err}"]

    for field in ("ticket", "sentiment", "category", "summary"):
        if not isinstance(data.get(field), str):
            problems.append(f"missing or non-string field: {field}")
    if problems:
        return problems

    if not (TICKETS_DIR / data["ticket"]).exists():
        problems.append(f"references unknown ticket: {data['ticket']}")
    if data["sentiment"] not in SENTIMENTS:
        problems.append(f'sentiment "{data["sentiment"]}" not in {sorted(SENTIMENTS)}')
    if data["category"] not in CATEGORIES:
        problems.append(f'category "{data["category"]}" not in {sorted(CATEGORIES)}')
    if not (SUMMARY_MIN <= len(data["summary"]) <= SUMMARY_MAX):
        problems.append(
            f"summary length {len(data['summary'])} outside {SUMMARY_MIN}-{SUMMARY_MAX}"
        )

    return problems


def main() -> int:
    files = sorted(OUTPUTS_DIR.glob("*.json"))
    failures = 0

    for path in files:
        problems = check_file(path)
        if not problems:
            print(f"  PASS  {path.name}")
        else:
            failures += 1
            print(f"  FAIL  {path.name}")
            for p in problems:
                print(f"        - {p}")

    print(f"\n{len(files) - failures}/{len(files)} outputs pass code checks")
    print("Remember: passing here only means structurally valid — not truthful.")
    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())
